### PR TITLE
Add timeout and improve message statistics

### PR DESCRIPTION
**Pull request by zehortigoza**
This PR depends on https://github.com/01org/mavlink-router/pull/15

### CODE REVIEW

File: endpoint.cpp [hunk 1]
```diff
@@ -95,7 +95,7 @@ Endpoint::~Endpoint()
     free(tx_buf.data);
 }
 
-int Endpoint::read_msg(struct buffer *pbuf, int *target_sysid, int *target_compid)
+int Endpoint::read_msg(struct buffer *pbuf, int *target_sysid)
```
Comment (lucasdemarchi) [inline endpoint.cpp]: > One system id can have several components ids.

In reality we may have multiple systemdid with multiple compid on the same Endpoint.  When doing the route we decide we would assume only one sysid + compid per endpoint.

Have you seem multiple compids in a single endpoint and this is hurting what you are doing here?  We may need to go the other direction and add support for more sysid/compid on a single endpoint rather than removing compid.

Comment (lucasdemarchi) [inline endpoint.cpp]: @edersondisouza ping

File: endpoint.cpp [hunk 2]
```diff
@@ -87,14 +89,27 @@ Endpoint::Endpoint(const char *name, bool crc_check_enabled)
...
+
+void Endpoint::handle_read()
+{
+    assert(_router);
+
+    int target_sysid;
+    struct buffer buf{};
+
+    if (read_msg(&buf, &target_sysid) > 0)
+        _router->router_msg(&buf, target_sysid, _system_id);
```
Comment (lucasdemarchi) [inline endpoint.cpp]: in the Endpoint we relied on read_msg() being called on a loop:

```
-    while (endpoint->read_msg(&buf, &target_sysid) > 0)
-        router_msg(endpoint, &buf, target_sysid);
-}
```

why did you change it to one read per loop iteration?

File: mainloop.cpp [hunk 3]
```diff
@@ -114,7 +114,7 @@ int Mainloop::write_msg(Endpoint *e, const struct buffer *buf)
     return r;
 }
 
-void Mainloop::router_msg(Endpoint *sender, struct buffer *buf, int target_sysid)
+void Mainloop::router_msg(struct buffer *buf, int target_sysid, int sender_sysid)
```
Comment (lucasdemarchi) [inline mainloop.cpp]: not from this commit, but this method should be called "route_msg"... i.e. a verb not a noun.

File: endpoint.cpp (hunk 2 above)
Comment (zehortigoza) [inline endpoint.cpp]: Ops my bad, I will go back again to while

File: mainloop.cpp (hunk 3 above)
Comment (zehortigoza) [inline mainloop.cpp]: yes it should, thanks

File: mainloop.h [hunk 4]
```diff
@@ -19,21 +19,22 @@
...
+#include "router.h"
 
 struct endpoint_entry {
     struct endpoint_entry *next;
     Endpoint *endpoint;
     bool remove;
 };
 
-class Mainloop {
+class Mainloop : public Router {
```
Comment (zehortigoza) [inline mainloop.h]: I don't think that we will need another Router implementation, Mainloop is only implementing Router because this way Endpoint can include router.h without cause a circular dependency.

**Comment by zehortigoza** [comment]
Fixed Lucas comments

**Comment by lucasdemarchi** [comment]
You can fix that by having some fwd declarations. No need to have Mainloop
implement the Router interface.

On Feb 8, 2017 4:51 PM, "José Roberto de Souza" <notifications@github.com>
wrote:

*@zehortigoza* commented on this pull request.
------------------------------

In mainloop.h <https://github.com/01org/mavlink-router/pull/16>:

>
 struct endpoint_entry {
     struct endpoint_entry *next;
     Endpoint *endpoint;
     bool remove;
 };

-class Mainloop {
+class Mainloop : public Router {

I don't think that we will need another Router implementation, Mainloop is
only implementing Router because this way Endpoint can include router.h
without cause a circular dependency.

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/01org/mavlink-router/pull/16>, or mute the thread

**Comment by lucasdemarchi** [comment]
Can you explain the numbers please?

```
                messages with CRC error: 916 0.163405% 48KBytes
                sequence of messages lost: 539945 96.320885%
                messages handled: 19708 792KBytes
```

So you counted 916 messages with crc error. But you are saying you had 539945 messages lost (due to resync, etc).... Given the seq number is uint8_t, it wraps quickly and you can only be sure of `256 * 916 == 234496`.... How come you concluded you have more than 2 times that maximum?

File: endpoint.cpp [hunk 5]
```diff
@@ -251,22 +246,48 @@ int Endpoint::read_msg(struct buffer *pbuf, int *target_sysid, int *target_compi
...
+
+            if (seq > _stat.read.expected_seq)
+                diff = (seq - _stat.read.expected_seq);
+            else
+                diff = (UINT8_MAX - _stat.read.expected_seq) + seq;
+
+            _stat.read.drop_seq_total += diff;
+            _stat.read.total += diff;
+        }
+        _stat.read.expected_seq = seq;
```
Comment (lucasdemarchi) [inline endpoint.cpp]: checking the math here.... I can't see any obvious thing that's wrong.  The final numbers don't sum up though.  Are you sure the other end is sending correct numbers?  Or do you have any explanation?

File: Makefile.am [hunk 6]
```diff
@@ -129,6 +129,9 @@ mavlink_routerd_SOURCES = \
 	main.cpp \
 	mainloop.cpp \
 	mainloop.h \
+	pollable.h \
```
Comment (otaviobp) [inline Makefile.am]: I am not a c++ expert, but shouldn't we use .hh or .hpp extensions?

File: pollable.cpp [hunk 7]
```diff
@@ -0,0 +1,26 @@
+/*
+ * This file is part of the MAVLink Router project
+ *
+ * Copyright (C) 2016  Intel Corporation. All rights reserved.
```
Comment (otaviobp) [inline pollable.cpp]: s/2016/2017/
The same for other headers.

**Comment by zehortigoza** [comment]
> Can you explain the numbers please?
>
>```
>                messages with CRC error: 916 0.163405% 48KBytes
>                sequence of messages lost: 539945 96.320885%
>                messages handled: 19708 792KBytes
>```
>
>So you counted 916 messages with crc error. But you are saying you had 539945 messages lost >(due to resync, etc).... Given the seq number is uint8_t, it wraps quickly and you can only be sure of >`256 * 916 == 234496`.... How come you concluded you have more than 2 times that maximum?

The calculation of sequence lost only takes place after CRC, so this `256 * 916 == 234496` do not apply.

File: Makefile.am (hunk 6 above)
Comment (lucasdemarchi) [inline Makefile.am]: no, please no

File: pollable.cpp (hunk 7 above)
Comment (zehortigoza) [inline pollable.cpp]: thanks

**Comment by zehortigoza** [comment]
Merged

**Comment by lucasdemarchi** [comment]
To clarify here for possible future reference. The issue with the math is that for each message sucessfully received we can have 255 with error .. so the maximum theoretical errored messages are n_success * 255 and not n_crc_errored
